### PR TITLE
The senvcfg.SSE will read (only) as zero when menvcfg.SSE is 0

### PR DESCRIPTION
**Pull request by chihminchao**

### CODE REVIEW

**Comment by aswaterman** [comment]
@ved-rivos can you approve?  Seems right to me, given the following:

https://github.com/riscv/riscv-isa-manual/blob/14a3c7965a7621b8ba7db9cc83f85e48ae3a0280/src/machine.adoc#L2208-L2209

File: riscv/csrs.h [hunk 1]
```diff
@@ -783,10 +785,19 @@ class sstateen_csr_t: public hstateen_csr_t {
...
+// senvcfg.sse is read_only 0 when menvcfg.sse = 0
 class senvcfg_csr_t final: public envcfg_csr_t {
  public:
-  senvcfg_csr_t(processor_t* const proc, const reg_t addr, const reg_t mask, const reg_t init);
+  senvcfg_csr_t(processor_t* const proc, const reg_t addr, const reg_t mask, const reg_t init, csr_t_p menvcfg);
+
+  reg_t read() const noexcept override {
+    return (menvcfg->read() | ~(MENVCFG_SSE)) & masked_csr_t::read();
+  }
+
```
Comment (ved-rivos) [inline riscv/csrs.h]: This seems to be missing the case where v=1, menvcfg.sse is 1 but henvcfg.sse is 0..
V=0: if menvcfg.sse is 0 then senvcfg.sse is read only zero.
V=1: if henvcfg.sse is 0 then senvcfg.sse is read only zero. (since henvcfg.sse is read only zero when menvcfg.sse is 0, only testing henvcfg.sse suffices).

**Comment by ved-rivos** [comment]
Looks good to me but seems to be missing the V=1 case.

**Comment by chihminchao** [comment]
@binno please comment

File: riscv/csrs.h (hunk 1 above)
Comment (binno) [reply riscv/csrs.h]: Yes, you're right. I missed the description for henvcfg.sse when v = 1. I will re-push my commits later.

**Comment by aswaterman** [comment]
What Ved wrote sounds right to me.

File: riscv/csrs.h (hunk 1 above)
Comment (chihminchao) [reply riscv/csrs.h]: fix in latest version

File: riscv/csrs.cc [hunk 2]
```diff
@@ -1561,6 +1555,17 @@ void senvcfg_csr_t::verify_permissions(insn_t insn, bool write) const {
   masked_csr_t::verify_permissions(insn, write);
 }
 
+// senvcfg.sse is read_only 0 when menvcfg.sse = 0
+reg_t senvcfg_csr_t::read() const noexcept {
+  return (((state->v)? state->henvcfg->read() : state->menvcfg->read()) |
+      (~MENVCFG_SSE)) & masked_csr_t::read();
```
Comment (ved-rivos) [inline riscv/csrs.cc]: @chihminchao Should this have been:
return (masked_csr_t::read() & ~((state->v ? state->henvcfg->read() : state->menvcfg->read()) & MENVCFG_SSE ? 0 : MENVCFG_SSE));

File: riscv/csrs.cc [hunk 3]
```diff
@@ -1570,8 +1575,18 @@ void henvcfg_csr_t::verify_permissions(insn_t insn, bool write) const {
...
+// henvcfg.dte is read_only 0 when menvcfg.dte = 0
+// henvcfg.sse is read_only 0 when menvcfg.sse = 0
+reg_t henvcfg_csr_t::read() const noexcept {
+  return (state->menvcfg->read() |
+      ~(MENVCFG_PBMTE | MENVCFG_STCE | MENVCFG_ADUE | MENVCFG_DTE | MENVCFG_SSE)) & masked_csr_t::read();
+}
+
 bool henvcfg_csr_t::unlogged_write(const reg_t val) noexcept {
-  const reg_t mask = menvcfg->read() | ~(MENVCFG_PBMTE | MENVCFG_STCE | MENVCFG_ADUE | MENVCFG_DTE);
+  const reg_t mask = state->menvcfg->read() | ~(MENVCFG_PBMTE | MENVCFG_STCE | MENVCFG_ADUE | MENVCFG_DTE);
```
Comment (ved-rivos) [inline riscv/csrs.cc]:
```suggestion
  const reg_t mask = state->menvcfg->read() | ~(MENVCFG_PBMTE | MENVCFG_STCE | MENVCFG_ADUE | MENVCFG_DTE | MENVCFG_SSE);
```

File: riscv/csrs.h [hunk 4]
```diff
@@ -475,25 +475,14 @@ class envcfg_csr_t: public masked_csr_t {
...
-// henvcfg.hade is read_only 0 when menvcfg.hade = 0
-// henvcfg.dte is read_only 0 when menvcfg.dte = 0
 class henvcfg_csr_t final: public envcfg_csr_t {
  public:
-  henvcfg_csr_t(processor_t* const proc, const reg_t addr, const reg_t mask, const reg_t init, csr_t_p menvcfg);
-
-  reg_t read() const noexcept override {
-    return (menvcfg->read() | ~(MENVCFG_PBMTE | MENVCFG_STCE | MENVCFG_ADUE | MENVCFG_DTE)) & masked_csr_t::read();
-  }
-
```
Comment (ved-rivos) [inline riscv/csrs.h:489]: Its unclear why these changes/restructuring was needed.

File: riscv/csrs.cc [hunk 5]
```diff
@@ -957,12 +957,6 @@ bool envcfg_csr_t::unlogged_write(const reg_t val) noexcept {
   return masked_csr_t::unlogged_write(adjusted_val);
 }
 
-// implement class henvcfg_csr_t
-henvcfg_csr_t::henvcfg_csr_t(processor_t* const proc, const reg_t addr, const reg_t mask, const reg_t init, csr_t_p menvcfg):
-  envcfg_csr_t(proc, addr, mask, init),
-  menvcfg(menvcfg) {
-}
-
```
Comment (ved-rivos) [inline riscv/csrs.cc]: Its unclear why this change/restructuring was needed.

File: riscv/csrs.cc (hunk 3 above)
Comment (binno) [reply riscv/csrs.cc]: Should we take care write operation ? specification only mention read status.

> The henvcfg.SSE and senvcfg.SSE fields will read as zero and are read-only.

File: riscv/csrs.cc (hunk 2 above)
Comment (binno) [reply riscv/csrs.cc]: I think the same behavior with henvcfg::read() is better.

File: riscv/csrs.h (hunk 4 above)
Comment (binno) [reply riscv/csrs.h:489]: I just only want to put implementation of member functions together.

**Comment by aswaterman** [comment]
I'll approve and merge once Ved's comments are addressed/responded to.

File: riscv/csrs.cc (hunk 5 above)
Comment (binno) [reply riscv/csrs.cc]: I just only want to put member functions with the same class together.

File: riscv/csrs.cc (hunk 3 above)
Comment (ved-rivos) [reply riscv/csrs.cc]: specification states "are read-only" i.e. not writable.

File: riscv/csrs.h (hunk 4 above)
Comment (ved-rivos) [reply riscv/csrs.h:489]: Since that restructuring is unrelated to this PR, I suggest that this PR not do this. The refactoring/restructuring of code should be in a separate PR.

File: riscv/csrs.cc (hunk 5 above)
Comment (ved-rivos) [reply riscv/csrs.cc]: Since that restructuring/refactoring is unrelated to this PR, I suggest that this PR not do this. The refactoring/restructuring of code should be in a separate PR.

File: riscv/csrs.cc (hunk 2 above)
Comment (ved-rivos) [reply riscv/csrs.cc]: I am not sure why its better to use the scope operator instead of derefencing the object.  Its also not clear to me why it is better to use scope operator for invoking read method of henvcfg class but dereferencing the menvcfg object is okay. I am not that versed with C++ nuances.

File: riscv/csrs.cc (hunk 5 above)
Comment (binno) [reply riscv/csrs.cc]: Sure

File: riscv/csrs.h (hunk 4 above)
Comment (binno) [reply riscv/csrs.h:489]: Sure

File: riscv/csrs.cc (hunk 2 above)
Comment (binno) [reply riscv/csrs.cc]: I agree your point, and I will revise it, thanks

File: riscv/csrs.cc (hunk 3 above)
Comment (binno) [reply riscv/csrs.cc]: Sure. I will revise it, thanks

**Comment by binno** [comment]
@ved-rivos , please help to review it again, thanks

File: riscv/csrs.h [hunk 6]
```diff
@@ -783,9 +782,11 @@ class sstateen_csr_t: public hstateen_csr_t {
   virtual bool unlogged_write(const reg_t val) noexcept override;
 };
 
+// senvcfg.sse is read_only 0 when menvcfg.sse = 0
```
Comment (ved-rivos) [inline riscv/csrs.h]: This comment should be placed before the read() method definition in csrs.cc.

File: riscv/csrs.cc [hunk 7]
```diff
@@ -1561,6 +1561,11 @@ void senvcfg_csr_t::verify_permissions(insn_t insn, bool write) const {
   masked_csr_t::verify_permissions(insn, write);
 }
 
+reg_t senvcfg_csr_t::read() const noexcept {
+  reg_t envcfg = (state->v) ? state->henvcfg->read() : state->menvcfg->read();
+  return masked_csr_t::read() & ~(envcfg & MENVCFG_SSE ? 0 : MENVCFG_SSE);
+}
+
```
Comment (ved-rivos) [inline riscv/csrs.cc:1569]: @binno Please also take care of the unlogged_write().

**Comment by ved-rivos** [comment]
@binno Thanks! I left two additional comments:
1. senvcfg::unlogged_write needs to be fixed.
2. The comment about senvcfg.sse in csrs.h would want to move to where the read method is defined in csrs.cc

**Comment by binno** [comment]
@ved-rivos , please help to review it again, thanks

**Comment by chihminchao** [comment]
@binno I think you need to resolve the reviewer's conversation on older change if the related issues have been fix in latest change

**Comment by ved-rivos** [comment]
@binno - Thanks. Looks good to me.

File: riscv/csrs.cc [hunk 8]
```diff
@@ -1561,6 +1561,18 @@ void senvcfg_csr_t::verify_permissions(insn_t insn, bool write) const {
   masked_csr_t::verify_permissions(insn, write);
 }
 
+// senvcfg.sse is read_only 0 when menvcfg.sse = 0
+reg_t senvcfg_csr_t::read() const noexcept {
+  reg_t envcfg = (state->v) ? state->henvcfg->read() : state->menvcfg->read();
```
Comment (ved-rivos) [inline riscv/csrs.cc]: @binno - one nit - (state->v)  can be just state->v. The parenthesis are not needed.

File: riscv/csrs.cc [hunk 9]
```diff
@@ -1561,6 +1561,18 @@ void senvcfg_csr_t::verify_permissions(insn_t insn, bool write) const {
...
+  return masked_csr_t::read() & ~(envcfg & MENVCFG_SSE ? 0 : MENVCFG_SSE);
+}
+
+bool senvcfg_csr_t::unlogged_write(const reg_t val) noexcept {
+  reg_t envcfg = (state->v) ? state->henvcfg->read() : state->menvcfg->read();
```
Comment (ved-rivos) [inline riscv/csrs.cc]: Same here about the parenthesis.

**Review by ved-rivos — APPROVED** [review]
Thanks!

**Review by aswaterman — APPROVED** [review]
Approving on the basis of @ved-rivos's review.
